hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/SimpleSchemaConflictResolutionStrategy.java

```diff
@@ -0,0 +1,178 @@
+import static org.apache.hudi.client.transaction.SchemaConflictResolutionStrategy.throwConcurrentSchemaEvolutionException;
+import static org.apache.hudi.common.table.timeline.HoodieTimeline.COMPACTION_ACTION;
+import static org.apache.hudi.common.table.timeline.HoodieTimeline.REPLACE_COMMIT_ACTION;
+import static org.apache.hudi.common.table.timeline.InstantComparison.LESSER_THAN_OR_EQUALS;
+import static org.apache.hudi.common.table.timeline.InstantComparison.compareTimestamps;
+
+/**
+ * The implementation of SchemaConflictResolutionStrategy that detects incompatible
+ * schema evolution from multiple writers
+ */
+public class SimpleSchemaConflictResolutionStrategy implements SchemaConflictResolutionStrategy {
+  private static final Logger LOG = LoggerFactory.getLogger(SimpleSchemaConflictResolutionStrategy.class);
+
+  @Override
+  public Option<Schema> resolveConcurrentSchemaEvolution(
+      HoodieTable table,
+      HoodieWriteConfig config,
+      Option<HoodieInstant> lastCompletedTxnOwnerInstant,
+      Option<HoodieInstant> currTxnOwnerInstant) {
+
+    // If this is compaction table service, skip schema evolution check as it does not evolve schema.
+    if (!currTxnOwnerInstant.isPresent()
+        || currTxnOwnerInstant.get().getAction().equals(COMPACTION_ACTION)
+        || (currTxnOwnerInstant.get().getAction().equals(REPLACE_COMMIT_ACTION)
+        && ClusteringUtils.isClusteringInstant(table.getMetaClient().getActiveTimeline(), currTxnOwnerInstant.get(), table.getMetaClient().getInstantGenerator()))) {
+      return Option.empty();
+    }
+
+    // Guard against unrecognized cases where writers do not come with a writer schema.
+    if (StringUtils.isNullOrEmpty(config.getWriteSchema())) {
+      LOG.warn(StringUtils.join("Writer config does not come with a valid writer schema. Writer config: ",
+          config.toString(), ". Owner instant: ", currTxnOwnerInstant.get().toString()));
+      return Option.empty();
+    }
+
+    Schema writerSchemaOfTxn = new Schema.Parser().parse(config.getWriteSchema());
+    // If a writer does not come with a meaningful schema, skip the schema resolution.
+    if (isSchemaNull(writerSchemaOfTxn)) {
+      return getTableSchemaAtInstant(new ConcurrentSchemaEvolutionTableSchemaGetter(table.getMetaClient()), currTxnOwnerInstant.get());
+    }
+
+    // Fast path: We can tell there is no schema conflict by just comparing the instants without involving table/writer schema comparison.
+    ConcurrentSchemaEvolutionTableSchemaGetter schemaResolver = new ConcurrentSchemaEvolutionTableSchemaGetter(table.getMetaClient());
+
+    // schema and writer schema.
+    HoodieInstant lastCompletedInstantAtTxnStart = lastCompletedTxnOwnerInstant.isPresent()
+        ? getInstantInTimelineImmediatelyPriorToTimestamp(
+        lastCompletedTxnOwnerInstant.get().requestedTime(), schemaResolver.computeSchemaEvolutionTimelineInReverseOrder()).orElse(null)
```

Comment:
Should this use completion time now?
```suggestion
        lastCompletedTxnOwnerInstant.get().getCompletionTime(), schemaResolver.computeSchemaEvolutionTimelineInReverseOrder()).orElse(null)
```

Reply:
done